hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/hdds/scm/TestCloseContainer.java

```diff
@@ -159,11 +182,98 @@ public void testCloseClosedContainer()
     StorageContainerManager scm = cluster.getStorageContainerManager();
     // Pick any container on the cluster and close it via client
     ContainerInfo container = scm.getContainerManager().getContainers().get(0);
+    // Checksum file doesn't exist before container close
+    List<HddsDatanodeService> hddsDatanodes = cluster.getHddsDatanodes();
+    for (HddsDatanodeService hddsDatanode: hddsDatanodes) {
+      assertFalse(containerChecksumFileExists(hddsDatanode, container));
+    }
+    // Close container
     OzoneTestUtils.closeContainer(scm, container);
+
+    // Checksum file exists after container close
+    for (HddsDatanodeService hddsDatanode: hddsDatanodes) {
+      GenericTestUtils.waitFor(() -> waitForContainerChecksumToExist(hddsDatanode, container), 100, 5000);
+    }
+
     assertThrows(IOException.class,
         () -> cluster.getStorageContainerLocationClient()
             .closeContainer(container.getContainerID()),
         "Container " + container.getContainerID() + " already closed");
   }
 
+  @Test
+  public void testContainerChecksumForClosedContainer() throws Exception {
+    // Create some keys to write data into the open containers
+    TestDataUtil.createKey(bucket, "key1", ReplicationFactor.THREE,
+        ReplicationType.RATIS, "this is the content");
+    StorageContainerManager scm = cluster.getStorageContainerManager();
+
+    ContainerInfo containerInfo1 = scm.getContainerManager().getContainers().get(0);
+    // Checksum file doesn't exist before container close
+    List<HddsDatanodeService> hddsDatanodes = cluster.getHddsDatanodes();
+    for (HddsDatanodeService hddsDatanode : hddsDatanodes) {
+      assertFalse(containerChecksumFileExists(hddsDatanode, containerInfo1));
+    }
+    // Close container.
+    OzoneTestUtils.closeContainer(scm, containerInfo1);
+    // Build expected merkle tree for one datanode.
+    HddsDatanodeService hddsDatanodeService = cluster.getHddsDatanodes().get(0);
+    OzoneContainer ozoneContainer = hddsDatanodeService.getDatanodeStateMachine().getContainer();
+    KeyValueContainer container1 = (KeyValueContainer) ozoneContainer.getController()
+        .getContainer(containerInfo1.getContainerID());
+    ContainerProtos.ContainerChecksumInfo expectedChecksumInfo1 = ContainerMerkleTreeTestUtils.readChecksumFile(
+        container1.getContainerData());
+
+    // Checksum file exists after container close and matches the expected container
+    // merkle tree for all the datanodes
+    for (HddsDatanodeService hddsDatanode : hddsDatanodes) {
+      GenericTestUtils.waitFor(() -> waitForContainerChecksumToExist(hddsDatanode, containerInfo1), 100, 5000);
+      container1 = (KeyValueContainer) ozoneContainer.getController().getContainer(containerInfo1.getContainerID());
```

Comment:
Can remove the cast, make `container1` have type `Container<?>`
```suggestion
      container1 = ozoneContainer.getController().getContainer(containerInfo1.getContainerID());
```